src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,225 @@
+
+
+def infer_cond(
+    applied_cond: itir.FunCall,
+    input_domain: SymbolicDomain | itir.FunCall,
+    offset_provider: Dict[str, Dimension],
+) -> Tuple[itir.FunCall, Dict[str, SymbolicDomain]]:
+    assert isinstance(applied_cond, itir.FunCall) and cpm.is_call_to(applied_cond, "cond")
+    assert (
+        isinstance(applied_cond.args[0], bool)
+        and isinstance(applied_cond.args[1], itir.FunCall)
+        and isinstance(applied_cond.args[2], itir.FunCall)
+    )
+    call_true, domains_true = infer_as_fieldop(applied_cond.args[1], input_domain, offset_provider)
+    call_false, domains_false = infer_as_fieldop(
+        applied_cond.args[2], input_domain, offset_provider
+    )
+    return im.cond(applied_cond.args[0], call_true, call_false), _merge_domains(
+        domains_true, domains_false
+    )
+
+
+def _validate_temporary_usage(body: list[itir.Stmt], temporaries: list[str]):
+    assigned_targets = set()
+    for stmt in body:
+        assert isinstance(stmt, itir.SetAt)  # TODO: extend for if-statements when they land
+        assert isinstance(
+            stmt.target, itir.SymRef
+        )  # TODO: stmt.target can be an expr, e.g. make_tuple
+        if stmt.target.id in assigned_targets:
+            raise ValueError("Temporaries can only be used once within a program.")
+        if stmt.target.id in temporaries:
+            assigned_targets.add(stmt.target.id)
+
+
+def infer_program(
+    program: itir.Program,
+    offset_provider: Dict[str, Dimension],
+) -> itir.Program:
+    accessed_domains: dict[str, SymbolicDomain] = {}
+    transformed_set_ats: list[itir.SetAt] = []
+
+    temporaries: list[str] = [tmp.id for tmp in program.declarations]
+
+    _validate_temporary_usage(program.body, temporaries)
+
+    for set_at in reversed(program.body):
+        assert isinstance(set_at, itir.SetAt)
+        if isinstance(set_at.expr, itir.SymRef):
+            transformed_set_ats.insert(0, set_at)
+            continue
+        assert isinstance(set_at.expr, itir.FunCall)
+        assert cpm.is_call_to(set_at.expr.fun, "as_fieldop")
+        assert isinstance(
+            set_at.target, itir.SymRef
+        )  # TODO: stmt.target can be an expr, e.g. make_tuple
+        if set_at.target.id in temporaries:
+            # ignore temporaries as their domain is the `AUTO_DOMAIN` placeholder
+            assert set_at.domain == AUTO_DOMAIN
+        else:
+            accessed_domains[set_at.target.id] = SymbolicDomain.from_expr(set_at.domain)
+
+        transformed_as_fieldop, current_accessed_domains = infer_as_fieldop(
+            set_at.expr, accessed_domains[set_at.target.id], offset_provider
+        )
+        transformed_set_ats.insert(
+            0,
+            itir.SetAt(
+                expr=transformed_as_fieldop,
+                domain=SymbolicDomain.as_expr(accessed_domains[set_at.target.id]),
+                target=set_at.target,
+            ),
+        )
```

Comment:
updating the domain should only be done if the original domain is `AUTO_DOMAIN`, correct? I am wondering if we should split this pass in 2. One that propagates domains between set_at (i.e. resolves AUTO_DOMAIN) and one that just annotates `as_fieldop`s. I find it confusing that we do both in the same pass, but not sure if I am missing something.

Reply:
I am also not sure if I am missing something, but I think this is only possible with some code duplication.
One option could be to add a pass that infers the domains of all `as_fieldop`s which is then only called by DaCe.
Let's discuss this next time we talk.

Reply:
As discussed: We discovered that running the domain propagation prior to temporary extraction should make all temporary related code in the propagation unnecessary and can hence be removed. We've placed a todo (in the let branch) and kept it in here for now so that we can move on.